
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Pipeline Overview &#8212; Psychcore NGS Germline Variant Pipeline  documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Getting Started" href="getting-started.html" />
    <link rel="prev" title="Contents" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="pipeline-overview">
<span id="sec-overview"></span><h1>Pipeline Overview<a class="headerlink" href="#pipeline-overview" title="Permalink to this headline">¶</a></h1>
<div class="section" id="genomics-workflow">
<h2>Genomics Workflow<a class="headerlink" href="#genomics-workflow" title="Permalink to this headline">¶</a></h2>
<p>The PsychCore NGS pipeline is a serverless, automated, easy to use genomics
pipeline for calling variants on large cohorts of human sequencing samples.
The workflow of the pipeline is as follows:</p>
<img alt="_images/PipelineWorkflowOverview.jpg" src="_images/PipelineWorkflowOverview.jpg" />
<p>The pipeline takes gzipped paired end sequencing fastq files (R1|R2.fastq.gz)
for each sample and aligns them to a specified reference build, using BWA MEM.
Several BAM processing steps (Picard and GATK) follow to produced a final
processed BAM which is haplotyped using Sentieon’s* Haplotyper module.
These steps are done in parallel for each sample in the cohort.
After Haplotyper is run, joint genotyping is performed accross the entire
cohort with Sentieon’s Genotyper, producing a VCF which undergoes VQSR (GATK)
which outputs the final VCF.</p>
</div>
<div class="section" id="pipeline-infrastructure">
<span id="infrastructure"></span><h2>Pipeline Infrastructure<a class="headerlink" href="#pipeline-infrastructure" title="Permalink to this headline">¶</a></h2>
<p><em>Stepfunctions, Batch, Lambda, Docker oh my!</em></p>
<p>Several AWS services are invovled in the infrastructure of the pipeline -
AWS <a class="reference external" href="https://aws.amazon.com/cloudformation/">Cloudformation</a>, <a class="reference external" href="https://aws.amazon.com/step-functions/">StepFunctions</a>, <a class="reference external" href="https://aws.amazon.com/lambda/">Lambda</a>, <a class="reference external" href="https://aws.amazon.com/batch/">Batch</a>, <a class="reference external" href="https://aws.amazon.com/ec2/">EC2</a>, <a class="reference external" href="https://aws.amazon.com/s3/">S3</a>, and <a class="reference external" href="https://aws.amazon.com/cloudwatch/">CloudWatch</a>.
The pipeline also makes use of <a class="reference external" href="https://www.docker.com">Docker</a> containers, as well as Google Cloud
Platform’s (<a class="reference external" href="https://cloud.google.com">GCP</a>) Google Cloud Storage (<a class="reference external" href="https://cloud.google.com/storage/">GCS</a>), and <a class="reference external" href="https://cloud.google.com/dataproc/">Dataproc</a>.</p>
<img alt="_images/PipelineInfrastructure.jpg" src="_images/PipelineInfrastructure.jpg" />
<p>The pipeline’s main bioinformatics tools are <a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA MEM</a>, <a class="reference external" href="https://broadinstitute.github.io/picard/">Picard</a>,
<a class="reference external" href="https://software.broadinstitute.org/gatk/documentation/tooldocs/3.8-0/">GATK</a>, and <a class="reference external" href="https://www.sentieon.com">Sentieon</a>.  Each of these tools has been (Docker)
containerised, and are deployed using AWS Batch. AWS Lambda
and AWS Stepfunctions manage the submission of jobs to the Batch
cluster as well as the handling of user input.  The entire system
architecture of the pipeline is managed by AWS Cloudformation while
logging during the run of the pipeline is handled by AWS Cloudwatch.</p>
<p><em>(Link to publication)</em></p>
</div>
<div class="section" id="a-note-about-sentieon">
<h2>A Note about Sentieon<a class="headerlink" href="#a-note-about-sentieon" title="Permalink to this headline">¶</a></h2>
<p>Sentieon developes and supplies a suite of bioinfromatics analyis tools for
processing genomics data.  In 2016, Sentieon won the PrecisionFDA <a class="reference external" href="https://precision.fda.gov/challenges/truth/results">Truth</a> and
<a class="reference external" href="https://precision.fda.gov/challenges/consistency/results">Consistency</a> challenges. Sentieon also won first place in
<a class="reference external" href="https://www.synapse.org/#!Synapse:syn312572/wiki/247695">ICGC TCGA Dream Mutation Calling Challenge</a>.
For more information see Sentieon’s homepage <a class="reference external" href="https://www.sentieon.com">here</a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Pipeline Overview</a><ul>
<li><a class="reference internal" href="#genomics-workflow">Genomics Workflow</a></li>
<li><a class="reference internal" href="#pipeline-infrastructure">Pipeline Infrastructure</a></li>
<li><a class="reference internal" href="#a-note-about-sentieon">A Note about Sentieon</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Contents</a></li>
      <li>Next: <a href="getting-started.html" title="next chapter">Getting Started</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/overview.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Michael Gilson, Claudia Dastmalchi, Lindsay Liang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/overview.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>